psycopg/psycopg/generators.py

```diff
@@ -251,28 +253,21 @@ def _pipeline_communicate(
 
 
 def _consume_notifies(pgconn: PGconn) -> None:
-    # Consume notifies
-    while True:
-        n = pgconn.notifies()
-        if not n:
-            break
-        if pgconn.notify_handler:
-            pgconn.notify_handler(n)
+    # Consume notifies. The lock ensures that notifies
+    # are consumed and handled atomically.
+    with lock:
```

Comment:
This is absolutely not ok. This lock is application-global and not even asyncio-friendly.

Why is a lock needed here anyway?

Reply:
The lock is there to satisfy that in `notifier`, if `wait(...)` completes then either notification processing has happened either:

1) in the `notifier` thread
2) in another thread

But not partially in one thread, partially in another – because this could mean that the `notifier` thread never actually gets to yield notifications but just goes back to sleep, waiting.

That said, you're absolute right that this global lock isn't workable. There's also the `with self.lock` around `wait(...)` that I'm a little unsure of – why is this one really necessary?

I don't consider it to be an implementation accident in the sense that I think `notifier` should work just like `add_notify_handler` which is as a kind of observable: as soon as the database sends the notification, we can pick it up either by registering a handler, or using `notifier` (or both).

But taking a step back, I think we could probably fix the code so that the implementation is solid, but at the same time, we should only take the time to do that if we can agree that it's the right behavior, worth committing to. I think we can avoid the lock that causes the delay you mention so that should probably not tip the boat.

Reply:
Just a quick note to say that `pgconn.consume_input` which calls into `PQconsumeInput` does not block:

```c
int
PQconsumeInput(PGconn *conn)
...
  /*
  * Load more data, if available. We do this no matter what state we are
  * in, since we are probably getting called because the application wants
  * to get rid of a read-select condition. Note that we will NOT block
  * waiting for more input.
  */
  if (pqReadData(conn) < 0)
    return 0;
```
(Reformatted to fit constraints here.)

That is, we can have two threads suspended waiting for the same socket to become read-ready and it's okay that both proceed to `pgconn.consume_input` in the sense that this won't cause any unexpected blocking.

Reply:
I think perhaps we should consider a `pgconn.consume_notifies() -> bool` instead – which then atomically (using a lock specific to this particular `pgconn`):

1. Retrieve a `PQnotifies`
2. Calls the `notify_handler` (if defined)

(Alternatively, the function could also do the `while True` loop, pulling out all notifications. In this case, the method doesn't have a return value).

The reason why it would seem we can't use the connection lock is that the `_consume_notifies` call is sort of intermingled between other blocks, since notifies are transmitted before results (apparently).

Reply:
I have reworked the implementation now, having better understood the requirements.

In `connection.py` I have added a _read lock_ which works to synchronize the program flow after a select on a file descriptor.

If the notifies thread holds the read lock, then an active query that resumes after a select that wants `READY_R` must wait to acquire this read lock – which is released as soon as the notifies thread has consumed the input and processed notifies.

Conversely, if the active query is able to acquire the read lock, the notifies thread simply waits for the query to complete before continuing its loop.

In the common case, where a connection is used exclusively by a single thread, there is virtually no cost: an uncontended lock is practically free to acquire and release.

As an implementation note for the asynchronous case, I did have to `socket.dup` the file descriptor inside `notifies()` in order for the asynchronous event loop to correctly fire (to achieve the same behavior as `select.select` where every blocked thread wakes up).